File: jaxlib/callback_custom_call.py
# Copyright 2021 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Python API for emitting a CustomCall op for host_callback.

The main entry point is `emit_custom_call`, called from `host_callback.py`
with the `callback_id` and the following operands:

 * an input token, to ensure proper op dependencies,
 * the current replica id of the calling device,
 * the actual arrays to pass to the callback, allocated on the calling device.

The `callback_id` and the shapes of the arguments and results are serializied
into a Descriptor bytearray using Flatbuffer (see callback_custom_call.fbs for
the buffer description, and callback_custom_call.cc for the encoding/decoding
logic.)

For CPU the descriptor array is passed as the very first operand (before the
input token), and for GPU it is passed as the `opaque` string.

The CPU implementation of the `callback_custom_call` is in
`callback_custom_call_cpu_py.cc` and for GPU it is in
`callback_custom_call_cuda_py.cc`. Those functions decode the Descriptor
argument, and eventually call `RunHostCallback` defined in
`callback_custom_call.cc`.
On the GPU, the input arrays are copied first to the host, and the results
are copied back to the device.

The actual invocation of the Python callback is done by `RunHostCallback`, by
way of a trampoline: a global function pointer set by the host_callback.py
upon initialization using the `set_callback_trampoline` entry point.
It is important for the Python code to set the trampoline to `None` when
exiting.

"""
from typing import Any, Callable, Optional, Tuple
import numpy as np

from jaxlib import xla_client

from . import callback_custom_call_cpu_py as callback_cpu  # type: ignore[import-error]
for _name, _value in callback_cpu.custom_call_registrations().items():
  xla_client.register_custom_call_target(_name, _value, platform="cpu")

try:
  from . import callback_custom_call_cuda_py as callback_cuda  # type: ignore[import-error]
  for _name, _value in callback_cuda.custom_call_registrations().items():
    xla_client.register_custom_call_target(_name, _value, platform="gpu")
except ImportError:
  pass  # We may have built without CUDA support

XlaOp = Any  # xla_extension.XlaOp
XlaShape = Any  # xla_client.Shape
XlaComputationBuilder = Any  # xla_bridge._JaxComputationBuilder


def set_callback_trampoline(trampoline: Optional[Callable]):
  callback_cpu.set_callback_trampoline(trampoline)


def emit_custom_call(platform: str, comp: XlaComputationBuilder,
                     callback_id: int, operands: Tuple[XlaOp, ...],
                     flat_results_xla_shapes: Tuple[XlaShape, ...], *,
                     ignore_results: bool) -> Tuple[XlaOp, ...]:
  """Emits the CustomCall to the given computation.

  Args:
    platform: the platform for which we are generating code, e.g., "cpu".
    comp: the computation into which to emit the ops.
    callback_id
    operands: a tuple containing: the input token, the replica id, and the
      arrays to pass to the callback.
    flat_results_xla_shapes: the expected shapes of results, including the
      output token at the start.
    ignore_results: whether the calling code ignores the callback results.
  """
  operand_shapes = tuple([
      _shape_with_default_layout(comp.GetShape(arg_op)) for arg_op in operands
  ])

  result_shapes = tuple([
      _shape_with_default_layout(res_shape)
      for res_shape in flat_results_xla_shapes
  ])

  descriptor_bytes = callback_cpu.encode_descriptor(
      callback_id, operand_shapes, ignore_results, result_shapes)

  if platform == "cpu":
    # Prepend the serialized descriptor.
    descriptor_op: XlaOp = xla_client.ops.Constant(
        comp, np.array(list(descriptor_bytes), np.uint8))
    operands = (descriptor_op,) + operands
    operand_shapes = (
        (_shape_with_default_layout(comp.GetShape(descriptor_op)),) +
        operand_shapes)
    opaque = b""
  elif platform == "gpu":
    opaque = bytes(descriptor_bytes)
  else:
    raise NotImplementedError(
        f"CustomCall callback not implemented for {platform}")

  call_results = xla_client.ops.CustomCallWithLayout(
      comp,
      b"callback_custom_call",
      operands=operands,
      shape_with_layout=xla_client.Shape.tuple_shape(result_shapes),
      operand_shapes_with_layout=operand_shapes,
      opaque=opaque,
      # The side_effect prevents the op from disappearing when there are no
      # results.
      has_side_effect=True)

  return tuple([
      xla_client.ops.GetTupleElement(call_results, i)
      for i in range(len(result_shapes))
  ])


def _shape_with_default_layout(xla_shape: XlaShape) -> XlaShape:
  if not xla_shape.is_array():
    return xla_client.Shape.token_shape()
  else:
    return xla_client.Shape.array_shape(
        xla_shape.element_type(), xla_shape.dimensions(),
        tuple(range(xla_shape.rank() - 1, -1, -1)))

File: jaxlib/callback_custom_call.cc
/* Copyright 2021 Google LLC

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "jaxlib/callback_custom_call.h"

#include <cstdint>
#import <iostream>
#include <sstream>
#include <vector>

#include "absl/container/flat_hash_map.h"
#include "absl/strings/numbers.h"
#include "absl/strings/str_format.h"
#include "absl/strings/str_split.h"
#include "flatbuffers/flatbuffers.h"
#include "jaxlib/callback_custom_call_generated.h"
#include "jaxlib/kernel_helpers.h"
#include "include/pybind11/detail/common.h"
#include "include/pybind11/pytypes.h"
#include "third_party/tensorflow/compiler/xla/layout_util.h"
#include "third_party/tensorflow/compiler/xla/literal.h"
#include "third_party/tensorflow/compiler/xla/python/types.h"
#include "third_party/tensorflow/compiler/xla/shape.h"
#include "third_party/tensorflow/compiler/xla/shape_util.h"
#include "third_party/tensorflow/compiler/xla/status_macros.h"
#include "third_party/tensorflow/compiler/xla/xla_data.proto.h"
#include "third_party/tensorflow/core/platform/logging.h"

namespace jax {

namespace py = pybind11;

namespace callback_custom_call_fb {

flatbuffers::Offset<flatbuffers::String> EncodeShape(
    flatbuffers::FlatBufferBuilder &builder, xla::Shape shape) {
  return builder.CreateString(shape.SerializeAsString());
}

xla::Shape DecodeShape(const flatbuffers::String *shape_fb) {
  xla::ShapeProto shape_proto;
  shape_proto.ParseFromString(flatbuffers::GetString(shape_fb));
  xla::Shape shape(shape_proto);
  return shape;
}

py::bytes EncodeDescriptor(uint32_t callback_id,
                           std::vector<xla::Shape> operand_shapes,
                           bool ignore_results,
                           std::vector<xla::Shape> result_shapes) {
  flatbuffers::FlatBufferBuilder builder(1024);
  // Serialize in reverse order
  std::vector<flatbuffers::Offset<flatbuffers::String>> operand_offset_vector;
  operand_offset_vector.resize(operand_shapes.size());
  for (int i = operand_shapes.size() - 1; i >= 0; --i) {
    operand_offset_vector[i] = EncodeShape(builder, operand_shapes[i]);
  }
  auto operands_encoded = builder.CreateVector(operand_offset_vector);

  std::vector<flatbuffers::Offset<flatbuffers::String>> result_offset_vector;
  result_offset_vector.resize(result_shapes.size());
  for (int i = result_shapes.size() - 1; i >= 0; --i) {
    result_offset_vector[i] = EncodeShape(builder, result_shapes[i]);
  }
  auto results_encoded = builder.CreateVector(result_offset_vector);
  auto descriptor = CreateCallbackDescriptor(
      builder, callback_id, operands_encoded, ignore_results, results_encoded);
  builder.Finish(descriptor);

  uint8_t *buffer = builder.GetBufferPointer();
  size_t size = builder.GetSize();
  return py::bytes(reinterpret_cast<char *>(buffer), size);
}

Descriptor DecodeDescriptor(const void *buffer) {
  const CallbackDescriptor *descr_fb = GetCallbackDescriptor(buffer);
  std::vector<xla::Shape> operands;
  for (auto op : *(descr_fb->operands())) {
    operands.push_back(DecodeShape(op));
  }
  std::vector<xla::Shape> results;
  for (auto res : *(descr_fb->results())) {
    results.push_back(DecodeShape(res));
  }
  return Descriptor{descr_fb->callback_id(), operands,
                    descr_fb->ignore_results(), results};
}

}  // namespace callback_custom_call_fb

// Store here the pointer to the Python function to call. The Python code should
// call SetCallbackTrampoline to set this to None before exiting.
static CallbackTrampolineToPython the_trampoline_ = nullptr;

void SetCallbackTrampoline(CallbackTrampolineToPython trampoline) {
  the_trampoline_ = trampoline;
}

std::vector<const void *> RunHostCallback(
    callback_custom_call_fb::Descriptor &descriptor,
    std::vector<const void *> arrays) {
  // descriptor.operands describes: token, arrays
  VLOG(2) << "Calling RunHostCallback with " << arrays.size() << " operands";
  py::gil_scoped_acquire gil_acquire;
  int nr_ops = descriptor.operands.size() - 1;
  CHECK_EQ(nr_ops, arrays.size());

  py::list ops_list;
  for (int i = 0; i < nr_ops; ++i) {
    xla::Shape arg_shape = descriptor.operands[1 + i];
    py::dtype dtype =
        xla::PrimitiveTypeToDtype(arg_shape.element_type()).ValueOrDie();
    VLOG(2) << "Preparing operand [" << i << "] of shape " << arg_shape.ToString();
    py::array arg_array(dtype, arg_shape.dimensions(),
                        xla::ByteStridesForShape(arg_shape),
                        const_cast<void *>(arrays[i]));
    ops_list.insert(i, arg_array);
  }
  py::object result = the_trampoline_(descriptor.callback_id, ops_list);

  int nr_results = descriptor.results.size() - 1;

  py::sequence result_sequence = py::cast<py::sequence>(result);
  if (! descriptor.ignore_results) {
    CHECK_EQ(nr_results, result_sequence.size());
  }
  std::vector<const void *> outputs;
  for (int i = 0; i < nr_results; ++i) {
    py::array result_array = result_sequence[i];
    outputs.push_back(result_array.data());
  }
  VLOG(2) << "Returning from RunHostCallback with " << outputs.size() << " results";
  return outputs;
}



}  // namespace jax
